File: python/two_bucket.py
from collections import deque


def measure(bucket_one, bucket_two, goal, start_bucket):
    q = deque()
    q.append((0, 0, 0))
    st = set()
    while q:
        (b1, b2, ct) = q.popleft()
        if (b1, b2) in st:
            continue
        if (start_bucket == 'one'):
            if b1 == 0 and b2 == bucket_two:
                continue
        else:
            if b1 == bucket_one and b2 == 0:
                continue
        st.add((b1, b2))
        if b1 == goal:
            return (ct, 'one', b2)
        if b2 == goal:
            return (ct, 'two', b1)

        # 1 pouring 1 to 2
        diff = min(bucket_two - b2, b1)
        q.append((b1 - diff, b2 + diff, ct + 1))
        # 2 emptying 1
        q.append((0, b2, ct + 1))
        # 3 filling 1
        q.append((bucket_one, b2, ct + 1))
        # 4 pouring 2 to 1
        diff = min(bucket_one - b1, b2)
        q.append((b1 + diff, b2 - diff, ct + 1))
        # 5 emptying 2
        q.append((b1, 0, ct + 1))
        # 6 filling 2
        q.append((b1, bucket_two, ct + 1))

    raise ValueError('can not measure')

File: dart/two_bucket.dart
import 'dart:collection';
import 'dart:math';

typedef Result = ({int moves, String goalBucket, int otherBucket});

class TwoBucket {
  final int bucketOne;
  final int bucketTwo;
  final int goal;
  final String startBucket;

  TwoBucket({
    required this.bucketOne,
    required this.bucketTwo,
    required this.goal,
    required this.startBucket,
  });

  Result measure() {
    final seen = <(int, int)>{};
    final q = Queue<(int, int, int)>();
    if (startBucket == "one") {
      q.add((bucketOne, 0, 1));
    } else {
      q.add((0, bucketTwo, 1));
    }

    while (q.isNotEmpty) {
      final (b1, b2, ct) = q.removeFirst();
      if (b1 == goal) {
        return (moves: ct, goalBucket: "one", otherBucket: b2);
      }

      if (b2 == goal) {
        return (moves: ct, goalBucket: "two", otherBucket: b1);
      }

      if (startBucket == "one") {
        if ((b1, b2) == (0, bucketTwo)) {
          continue;
        }
      } else {
        if ((b1, b2) == (bucketOne, 0)) {
          continue;
        }
      }

      if (seen.contains((b1, b2))) {
        continue;
      }

      seen.add((b1, b2));

      int p = min(bucketTwo - b2, b1);
      q.addLast((b1 - p, b2 + p, ct + 1));
      q.addLast((0, b2, ct + 1));
      q.addLast((bucketOne, b2, ct + 1));

      p = min(bucketOne - b1, b2);
      q.addLast((b1 + p, b2 - p, ct + 1));
      q.addLast((b1, 0, ct + 1));
      q.addLast((b1, bucketTwo, ct + 1));
    }

    throw ArgumentError("impossible");
  }
}
